firebase/functions/src/payments.ts

```diff
@@ -0,0 +1,183 @@
+import * as functions from "firebase-functions";
+import * as admin from "firebase-admin";
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires,max-len
+const stripe = require("stripe")(functions.config().clothingloop.stripeKey);
+const endpointSecret = functions.config().clothingloop.webhookSecret;
+
+interface IPaymentInitiateData {
+  variables: {
+    amount: number | null;
+    email: string;
+    type: string;
+    priceId: string;
+  }
+}
+
+const payments = {
+  initiate: async (data : IPaymentInitiateData): Promise<{ sessionId: string }> => {
+    functions.logger.debug("paymentInitiate parameters", data);
+    const {amount, email, type, priceId} = data.variables;
+
+    let options : any = {
+      payment_method_types: ["ideal", "card"],
+      line_items: [
+        {
+          price_data: {
+            currency: "eur",
+            product_data: {
+              name: "Donation",
+            },
+            unit_amount: amount,
+          },
+          quantity: 1,
+        },
+      ],
+      mode: "payment",
+      success_url: functions.config().clothingloop.base_domain + "/donate/thankyou",
+      cancel_url: functions.config().clothingloop.base_domain + "/donate/cancel",
+    };
+
+    if (type === "recurring") {
+      options = {
+        payment_method_types: ["sepa_debit", "card"],
+        mode: "setup",
+        success_url: functions.config().clothingloop.base_domain + "/donate/thankyou",
+        cancel_url: functions.config().clothingloop.base_domain + "/donate/cancel",
+        metadata: {
+          price_id: priceId,
+        },
+      };
+    }
+
+    functions.logger.debug(options);
+    if (email && email !== "") {
+      options.customer_email = email;
+    }
+
+    const session = await stripe.checkout.sessions.create(options).catch((error: { type: string, message: string }) => {
+      functions.logger.warn(`Error from Stripe: ${error.type}: ${error.message}`);
+      throw new functions.https.HttpsError("unknown", "Something went wrong when processing your checkout request...");
+    });
+
+    if (!session || !session.id) {
+      throw new functions.https.HttpsError("unknown", "Something went wrong when processing your checkout request...");
+    }
+    await admin.firestore().collection("payments").doc(session.id).set({
+      sessionId: session.id,
+      createdAt: admin.firestore.FieldValue.serverTimestamp(),
+      amount,
+      email,
+      recurring: type === "recurring",
+    });
+
+    functions.logger.debug("payment initiate result", session);
+    return {
+      sessionId: session && session.id,
+    };
+  },
+  webhook: async (request: functions.Request, response: functions.Response) : Promise<any> => {
+    functions.logger.debug("payment webhook parameters", request.body);
+
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
```

Comment:
Why is it necessary to disable types several times? Doesn't look like you're doing anything especially "illegal". Casting to any might be simpler in this case.

Reply:
No idea, couldn't make TypeScript happy otherwise